### ISSUE TITLE
instance_set_transform: Condition ' Math::is_nan(v.x) ' is true

### ISSUE THREAD

**Issue by zappedfish** [godotengine/godot#50450]
### Godot version

3.3 Stable

### System information

GLES3  ||  Windows 7 NVIDIA / Windows 10 NVIDIA 2070 SUPER / Fedora 34 AMD Radeon R7 M340

### Issue description

Sometimes my game hangs and starts spamming` instance_set_transform: Condition ' Math::is_nan(v.x) ' is true` to the console. When that happens, the game stops rendering and outputs just a plain color to the screen only. 

1. **Windows 7 NVIDIA** 
    Happened very often. Happened even when the game was idling.
2. **Windows 10 NVIDIA 2070 SUPER** 
    Happened somewhat often. Happened when the game removed and created the camera again on a scene change.
3. **Fedora 34 AMD Radeon R7 M340**
    Happened rarely, when the game removed and created the camera again on a scene change.

Additionally:
- I used [Godot 360](https://github.com/Cykyrios/Godot360) in this project.
- Bug happened more frequently when I enabled multiple threads in the project settings.
- Never happened on Windows 7 with Intel HD Graphics 5500

I strongly suspect that it's related to the camera translation code. I think it stopped occurring when I removed everything that was related to the translation in it. Also, there is code that is translating the camera on level changes that may trigger the bug. (not the code itself, but the act of translation)

**_I apologize for uploading a complete 180MB project instead of a minimal reproduction one, I can't find time to make a minimal project..._**  

### Steps to reproduce

1. Skip the intro by pressing Z multiple times.
2. After that either just wait 10-20 minutes or...
2.1 Make a character enter other location. (just go up and you will stumble upon a door) You may need to re-enter that location multiple times (five, ten or maybe more)

Eventually the bug will occur. If it does not occur, enable multiple threads in the project settings and try again.

### Minimal reproduction project

https://drive.google.com/file/d/1Co-h-YtVqOETfqg5HVUlPmYC-4Buqj8O/view?usp=sharing

**Comment by beegor** [comment]
Also experiencing this bug and it's really hard to debug. It would be a great help if the error log message would contain the path of the problematic node or any other info which could help identify what object or part of GDScript caused the problem.
Currently, it's not possible to say if this happens due to a bug in the physics engine or is caused by something in gdscript.

**Comment by Klaim** [comment]
We are having the same issue with this [small amateur project](https://github.com/gamkedo-la/the-house), whatever the platform (though we target mainly HTML5).
The player can "take" items in their hand, in which case the item will be tracking the hand position and try to move there through velocity tweaking (similar to how steering behavior would have been, but we don´t have a general solution for that so went for the simplest). When moving close to another block (as in the "box" provided by default by godot) we end up with the item being stuck in some corner of blocks and the logs showing the same is_nan failure.
The player moves through `move_and_slide`, I don't know if thatś important though.

The player movement code: https://github.com/gamkedo-la/the-house/blob/main/player/player.gd#L66-L87

The code moving the item in the hand, which tries to track a specific spatial node: https://github.com/gamkedo-la/the-house/blob/main/items/item.gd#L110-L124

We have been seeing the issue since the beginning of the project as taking an object was implemented very early. Even when taking was implemented by just moving the item's node into the hierarchy of the player, we would hit the issue. So I suspect it have to do with how `move_and_slide` is implemented or used?

**Comment by theraot** [comment]
I cannot give you my code at this point, and this quite hard to reproduce with my project anyway… But I assure you `move_and_slide_with_snap` can return `Vector3(nan, nan, nan)`, and I believe `move_and_slide` can too.

You can check like this:

```
var new_velocity := move_and_slide(velocity)
if new_velocity != new_velocity:
    breakpoint

velocity = new_velocity
```

This breaks on `nan` since `nan != nan` is `true`.

**Comment by jbritain** [comment]
Was a solution for this ever found? My project which worked fine beforehand spontaneously started throwing this error every time I start it, making it completely unusable.

**Comment by Klaim** [comment]
This is still among the issues I keep having in my project (source above).

**Comment by Klaim** [comment]
So far the only workaround that worked for me was to release the game in Debug :/

**Comment by beegor** [comment]
> So far the only workaround that worked for me was to release the game in Debug :/
> I suspect that there are uninitialized variables at play in the C++ code of Godot.

I doubt it's a bug in Godots C++ code, it's more likely that you are passing NaN to it somewhere from your code. In my case, there was a part of the code that recursively increased the x coordinate until it exceeded the max int value. From the error, it was not clear which part of our gd script code caused the error, I found it by analyzing all the recent changes, and it was still not obvious. I 'v been lucky to find it and fix it. 

The message for this type of error should be more precise.

**Comment by Klaim** [comment]
> I doubt it's a bug in Godots C++ code, it's more likely that you are passing NaN to it somewhere from your code. 

While that would be my initial thought too, does this reasoning still holds if Debug mode does not expose the issue at all though? Because this is what I'm observing in my project.
I dont know enough yet of how Godot is implemented to understand the issue either. I intend to maybe try to build a debug version of godot and see for myself (I'm more versed in C++ than Godot/GDScript) If Godot does something to floats calculations in Debug, that might mean you are correct, I'll try to check more potential sources.

Also note that my project's code is readable there: https://github.com/gamkedo-la/the-house
See the code using the move functions there: https://github.com/gamkedo-la/the-house/blob/main/player/player.gd#L200-L233 (current version also have the workaround specified few messages before, but it does not work).

In this code `speed` is basically a runtime constant (depending on walking, crouching etc.), `movement_translation` is built using vector3 constants normalized, oriented movement is just `movement_translation` rotated through a valid quaternion (because it comes from Godot), gravity is litterally the value from the editor 9.x something, and we pass oriented_movement directly to the move functions. So in this speicfic code I'm not seeing how NaN could come in, there is no additions. But I might be indeed missing something.


To reproduce the issue:
- build a Release version of the game
- "Start" the game, walk around
- If you take an object (left click on one of the mushrooms for example) you might hit the issue faster
When the issue hits, you will get major performance drop, 1fps slideshow.

I have many mitigation in the code of the movement of the items: https://github.com/gamkedo-la/the-house/blob/main/items/interactive_item.gd#L159-L201
Basically when held by the player, they just follow the hand's position, in the same fashion as simple "arrive" steering behavior would do.
The issue seems to be detected in this code more often, but basically this is just reducing the distance between the object's position and the hand's position, so not sure how NaN get introduced here. Here `base_linear_velocity` comes from the code previously linked.
I'll have to investigate more I guess.

I'm interested if you find anything in this code that would point the issue.
